### PR TITLE
@W-9590264 Changes to convert Arrow Stream to result set.  

### PR DESCRIPTION
**Pull request by palaniappa**

### CODE REVIEW

File: src/main/java/com/salesforce/cdp/queryservice/util/ArrowUtil.java [hunk 1]
```diff
@@ -0,0 +1,105 @@
...
+			if(fieldVectors != null){
+				for(FieldVector fieldVector : fieldVectors) {
+					if(fieldVector !=  null) {
+						fieldVector.close();
+					}
+				}
+				fieldVectors = null;
+			}
+			allocator.close();
+			allocator =  null;
```
Comment (g1-pai) [inline src/main/java/com/salesforce/cdp/queryservice/util/ArrowUtil.java:79]: nit- do we need to set this explicitly?

File: src/main/java/com/salesforce/cdp/queryservice/util/ArrowUtil.java [hunk 2]
```diff
@@ -0,0 +1,105 @@
...
+				return value.toString();
+			return null;
+		}
+		else if(type == Types.MinorType.INT) {
+			return ((IntVector)fieldVector).get(index);
+		}
+		else if(type == Types.MinorType.DECIMAL) {
+			return ((DecimalVector)fieldVector).getObject(index);
+		}
+		throw new SQLException("Unknown arrow type {}", type.name());
```
Comment (g1-pai) [inline src/main/java/com/salesforce/cdp/queryservice/util/ArrowUtil.java:103]: Do we need to handle date time?

**Review by g1-pai — CHANGES_REQUESTED** [review]

**Review by g1-pai — COMMENTED** [review]
Thanks for making this change. Could you please upgrade pom version.

File: src/main/java/com/salesforce/cdp/queryservice/core/QueryServiceAbstractStatement.java [hunk 3]
```diff
@@ -61,7 +72,7 @@ public ResultSet executeQuery(String sql) throws SQLException {
         try {
             this.sql = sql;
             boolean isTableauQuery = isTableauQuery();
-            Response response = queryExecutor.executeQuery(sql, isTableauQuery ? Optional.of(Constants.MAX_LIMIT) : Optional.empty(), Optional.of(offset), isTableauQuery ? Optional.of("1 ASC") : Optional.empty());
+            Response response = queryExecutor.executeQuery(sql, this.connection.getEnableArrowStream(), isTableauQuery ? Optional.of(Constants.MAX_LIMIT) : Optional.empty(), Optional.of(offset), isTableauQuery ? Optional.of("1 ASC") : Optional.empty());
```
[the remark this answers is not in the document]
Comment (palaniappa) [reply src/main/java/com/salesforce/cdp/queryservice/core/QueryServiceAbstractStatement.java:75]: sure, will change that.

File: src/main/java/com/salesforce/cdp/queryservice/core/QueryServiceConnection.java [hunk 4]
```diff
@@ -35,12 +35,18 @@
...
     private String serviceRootUrl;
     private Token token;
+    private boolean enableArrowStream = false;
 
     public QueryServiceConnection(String url, Properties properties) {
         this.properties = properties;
         this.serviceRootUrl = url.substring(Constants.DATASOURCE_TYPE.length());
         this.properties.put(Constants.LOGIN_URL, serviceRootUrl);
         setupDefaultClientSecretsIfRequired(serviceRootUrl, this.properties);
+        if(this.properties.containsKey(Constants.ENABLE_ARROW_STREAM)) {
```
[the remark this answers is not in the document]
Comment (palaniappa) [reply src/main/java/com/salesforce/cdp/queryservice/core/QueryServiceConnection.java:45]: get returns an object and requires casting. So it isn't working out.

File: src/main/java/com/salesforce/cdp/queryservice/util/ArrowUtil.java [hunk 5]
```diff
@@ -0,0 +1,105 @@
...
+		byte[] bytes = Base64.getDecoder().decode(queryServiceResponse.getArrowStream());
+		ByteArrayInputStream inputStream = new ByteArrayInputStream(bytes);
+		List<FieldVector> fieldVectors = null;
+		RootAllocator allocator = new RootAllocator(Long.MAX_VALUE);
+		List<Map<String, Object>> data = new ArrayList<>();
+		try (ArrowStreamReader arrowStreamReader = new ArrowStreamReader(inputStream, allocator)) {
+
+			VectorSchemaRoot root = arrowStreamReader.getVectorSchemaRoot();
+			while (arrowStreamReader.loadNextBatch()) {
+				fieldVectors = root.getFieldVectors();
```
[the remark this answers is not in the document]
Comment (palaniappa) [reply src/main/java/com/salesforce/cdp/queryservice/util/ArrowUtil.java:44]: moved it outside the loop.

File: src/main/java/com/salesforce/cdp/queryservice/util/ArrowUtil.java (hunk 1 above)
Comment (palaniappa) [reply src/main/java/com/salesforce/cdp/queryservice/util/ArrowUtil.java:79]: not required.

File: src/main/java/com/salesforce/cdp/queryservice/util/ArrowUtil.java (hunk 2 above)
Comment (palaniappa) [reply src/main/java/com/salesforce/cdp/queryservice/util/ArrowUtil.java:103]: we are using varchar for date time as well. So we are good.

File: src/main/java/com/salesforce/cdp/queryservice/util/ArrowUtil.java [hunk 6]
```diff
@@ -0,0 +1,105 @@
...
+				int rowCount = fieldVectors.get(0).getValueCount();
+				for(int i=0;i<rowCount;++i) {
```
[the remark this answers is not in the document]
Comment (palaniappa) [reply src/main/java/com/salesforce/cdp/queryservice/util/ArrowUtil.java:46]: I did not get this.

[the remark this answers is not in the document]
Comment (g1-pai) [reply src/main/java/com/salesforce/cdp/queryservice/util/ArrowUtil.java:47]: In the response object also we have a rowcount field. I was thinking if we need an assert of rowcount field and rowcount from arrow.

**Review by g1-pai — APPROVED** [review]

**Pull request merged**